autowarefoundation/autoware.universe
# feat(default_ad_api): add operation mode api

Signed-off-by: Takagi, Isamu <isamu.takagi@tier4.jp>

## Description

This is an implementation of [operation mode API](https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-interfaces/ad-api/list/api/operation_mode/). For details, see [the discussion page](https://github.com/autowarefoundation/autoware/discussions/2464#discussioncomment-4107546).

## Related links

https://github.com/autowarefoundation/autoware.universe/pull/1535
https://github.com/autowarefoundation/autoware.universe/pull/1586
https://github.com/tier4/tier4_autoware_msgs/pull/46

## Tests performed

1. Launch the planning_simulator.launch.xml, initialize pose and set route.
2. Call `change_to_autonomous` service, and check if the vehicle starts.
3. Call `change_to_stop` service, and check if the vehicle stops.
4. Call `change_to_local` and `change_to_remote` services, and check if gate_mode and selector_mode are changed (AutowareStatePanel is useful).

## Notes for reviewers

Commands
`ros2 service call /api/operation_mode/change_to_stop autoware_adapi_v1_msgs/srv/ChangeOperationMode`
`ros2 service call /api/operation_mode/change_to_autonomous autoware_adapi_v1_msgs/srv/ChangeOperationMode`
`ros2 service call /api/operation_mode/change_to_local autoware_adapi_v1_msgs/srv/ChangeOperationMode`
`ros2 service call /api/operation_mode/change_to_remote autoware_adapi_v1_msgs/srv/ChangeOperationMode`

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
